File: frontend/src/contracts/task/schema.ts
/**
 * What a task is: a graph of triggers and agents, and the file it lives in. The shape only —
 * reading and writing one is the codec's, walking one is the graph's.
 */

import { extensionOf } from '@/path'
import { snap } from '../grid'

export const TASK_EXTENSION = '.task'

export function isTaskPath(path: string): boolean {
  return extensionOf(path) === 'task'
}

export type TaskKind = TaskNode['kind']

interface NodeBase {
  id: string
  name: string
  x: number
  y: number
  /** Switched off: the node does no work of its own and stands as a wire instead — what
   *  feeds it goes straight on to what it feeds. A trigger switched off never fires, which
   *  is how a task keeps its schedule on paper while you work on it. Absent is on, so
   *  every task written before the switch reads as one with everything on. */
  off?: true
}

/** The trigger that is you: the task runs when you press play, and never on its own. Every
 *  task is born with one, so there is always a way to run it by hand. */
export interface ManualTrigger extends NodeBase {
  kind: 'trigger.manual'
}

export interface IntervalTrigger extends NodeBase {
  kind: 'trigger.interval'
  minutes: number
}

export interface TimeTrigger extends NodeBase {
  kind: 'trigger.time'
  /** Local time of day, HH:MM. */
  at: string
}

export interface FileTrigger extends NodeBase {
  kind: 'trigger.file'
  /** The file watched, absolute or relative to the checkout the task lives in. */
  path: string
}

export interface ClaudeNode extends NodeBase {
  kind: 'agent.claude'
  prompt: string
  /** Name of a project persona whose PERSONA.md joins the agent's system prompt. */
  persona?: string
  /** How long the errand may take, in minutes. Unset is 5 — a step, not a day. */
  minutes?: number
}

export interface MuseNode extends NodeBase {
  kind: 'agent.muse'
  prompt: string
  /** Name of a project persona whose PERSONA.md joins the agent's system prompt. */
  persona?: string
  /** How long the errand may take, in minutes. Unset is 5 — a step, not a day. */
  minutes?: number
}

export interface ShellNode extends NodeBase {
  kind: 'agent.shell'
  /** Run by `sh -c` in the checkout, upstream output on stdin, stdout onward. */
  command: string
  minutes?: number
}

export interface GateNode extends NodeBase {
  kind: 'agent.gate'
  /** The branch continues only when the input matches this pattern — how "act on what
   *  the agent flagged" is written: the agent starts its answer with a word, the gate
   *  looks for it, and a quiet answer ends the branch without ending the run. */
  pattern: string
}

export interface NoteNode extends NodeBase {
  kind: 'agent.note'
  /** Project path of the note the run's output lands in. */
  path: string
  /** Add to the end instead of rewriting — how a recurring task keeps a log. */
  append?: boolean
}

/**
 * What every node that talks to GitHub has: which repository, and — where it watches one —
 * how long to leave it alone between looks. The repository is unset far more often than
 * not: a task lives in a checkout, the checkout has a remote, and the remote is the answer.
 */
interface GithubBase extends NodeBase {
  /** `owner/name`. Unset is the remote of the checkout the task lives in. */
  repo?: string
}

interface GithubWatchBase extends GithubBase {
  /** How long to leave GitHub alone between looks. Unset is 5 — the task beats every 30
   *  seconds, which is a fine rate for a file on this disk and a rude one for an API. */
  minutes?: number
}

export interface GithubIssueTrigger extends GithubWatchBase {
  kind: 'trigger.github.issue'
  /** A GitHub search, where one is wanted: `label:bug`, `author:someone`. Unset watches
   *  every issue in the repository. */
  query?: string
}

export interface GithubPullTrigger extends GithubWatchBase {
  kind: 'trigger.github.pull'
  /** As above — `review-requested:@me` is what this is for. */
  query?: string
}

export interface GithubMentionTrigger extends GithubWatchBase {
  kind: 'trigger.github.mention'
}

export interface GithubCheckTrigger extends GithubWatchBase {
  kind: 'trigger.github.check'
  /** The branch whose checks are watched. Unset is the branch the checkout is on. */
  branch?: string
}

export interface GithubCommentNode extends GithubBase {
  kind: 'agent.github.comment'
  /** The issue or pull request commented on. Unset is whatever the trigger was about. */
  number?: number
}

export interface GithubPullNode extends GithubBase {
  kind: 'agent.github.pull'
  /** What it is opened against. Unset is the repository's default branch. */
  base?: string
  /** What it is opened from. Unset is the branch the checkout is on. */
  head?: string
  /** The first line of the step's input is the title unless one is written here. */
  title?: string
  draft?: boolean
}

export type TaskNode =
  | ManualTrigger
  | IntervalTrigger
  | TimeTrigger
  | FileTrigger
  | GithubIssueTrigger
  | GithubPullTrigger
  | GithubMentionTrigger
  | GithubCheckTrigger
  | ClaudeNode
  | MuseNode
  | ShellNode
  | GithubCommentNode
  | GithubPullNode
  | GateNode
  | NoteNode

/** Everything that reaches GitHub, and the two questions worth asking of a node: whether
 *  it is one of them, and whether it is one of the ones that watches. Predicates rather
 *  than a string test, so a field only some of them carry is still typed. */
export type GithubWatchNode =
  | GithubIssueTrigger
  | GithubPullTrigger
  | GithubMentionTrigger
  | GithubCheckTrigger

export type GithubNode = GithubWatchNode | GithubCommentNode | GithubPullNode

export interface TaskEdge {
  from: string
  to: string
}

export interface Task {
  version: 1
  nodes: TaskNode[]
  edges: TaskEdge[]
}

export function isTrigger(node: TaskNode): boolean {
  return node.kind.startsWith('trigger.')
}

/** Whether a trigger will actually fire the task: wired into the graph, and switched on.
 *  Everything that keeps time or watches a source asks this one question. */
export function fires(node: TaskNode, wired: ReadonlySet<string>): boolean {
  return wired.has(node.id) && !node.off
}

/** How a trigger reads in a sentence — the tasks page's word for it. Null for agents. */
export function triggerLabel(node: TaskNode): string | null {
  switch (node.kind) {
    case 'trigger.manual':
      return 'triggered manually'
    case 'trigger.interval':
      return `every ${node.minutes} minute${node.minutes === 1 ? '' : 's'}`
    case 'trigger.time':
      return `at ${node.at}`
    case 'trigger.file':
      return `when ${node.path} changes`
    case 'trigger.github.issue':
      return `when an issue changes in ${node.repo ?? 'this repo'}${asked(node.query)}`
    case 'trigger.github.pull':
      return `when a pull request changes in ${node.repo ?? 'this repo'}${asked(
        node.query,
      )}`
    case 'trigger.github.mention':
      return 'when you are mentioned on GitHub'
    case 'trigger.github.check':
      return `when checks change on ${node.branch ?? 'this branch'}`
    default:
      return null
  }
}

/** The search a watch narrows itself with, where it has one — read as part of the sentence
 *  the trigger is, since a trigger watching one label is not watching the repository. */
function asked(query: string | undefined): string {
  return query ? ` matching ${query}` : ''
}

export function emptyTask(): Task {
  return {
    version: 1,
    nodes: [
      { id: 'trigger', kind: 'trigger.manual', name: 'Trigger manually', x: 80, y: 120 },
    ],
    edges: [],
  }
}

/** Every kind a task can hold, in the order the add menu offers them. */
export const TASK_KINDS: TaskKind[] = [
  'trigger.manual',
  'trigger.interval',
  'trigger.time',
  'trigger.file',
  'trigger.github.issue',
  'trigger.github.pull',
  'trigger.github.mention',
  'trigger.github.check',
  'agent.claude',
  'agent.muse',
  'agent.shell',
  'agent.github.comment',
  'agent.github.pull',
  'agent.gate',
  'agent.note',
]

/** The kinds that reach GitHub: a branch of their own in the add menu, under whichever
 *  family they belong to, and the ones the runtime needs a connection for. */
export const GITHUB_KINDS: TaskKind[] = TASK_KINDS.filter((kind) =>
  kind.includes('.github.'),
)

export function isGithub(kind: TaskKind): boolean {
  return kind.includes('.github.')
}

export function isGithubNode(node: TaskNode): node is GithubNode {
  return isGithub(node.kind)
}

export function isGithubWatch(node: TaskNode): node is GithubWatchNode {
  return isGithub(node.kind) && node.kind.startsWith('trigger.')
}

/** The kinds the editor no longer offers. A task written when it did still opens and runs;
 *  the menu just does not make any more of them. */
export const LEGACY_KINDS: TaskKind[] = ['agent.gate', 'agent.note']

/** What each kind is called: the word in the add menu, and the name a node of it wears
 *  until it is given another. */
export const KIND_LABEL: Record<TaskKind, string> = {
  'trigger.manual': 'Trigger manually',
  'trigger.interval': 'Every N minutes',
  'trigger.time': 'At a time',
  'trigger.file': 'When a file changes',
  'agent.claude': 'Claude Code',
  'agent.muse': 'Muse Code',
  'agent.shell': 'Command',
  'trigger.github.issue': 'When an issue changes',
  'trigger.github.pull': 'When a pull request changes',
  'trigger.github.mention': 'When you are mentioned',
  'trigger.github.check': 'When checks change',
  'agent.github.comment': 'Comment on GitHub',
  'agent.github.pull': 'Open a pull request',
  'agent.gate': 'Only if',
  'agent.note': 'Write a note',
}

/** The fields a node of each kind is born with, beyond the ones every node has. */
export const KIND_SEED: Record<TaskKind, Partial<TaskNode>> = {
  'trigger.manual': {},
  'trigger.interval': { minutes: 30 },
  'trigger.time': { at: '09:00' },
  'trigger.file': { path: '' },
  'agent.claude': { prompt: '' },
  'agent.muse': { prompt: '' },
  'agent.shell': { command: '' },
  'trigger.github.issue': {},
  'trigger.github.pull': {},
  'trigger.github.mention': {},
  'trigger.github.check': {},
  'agent.github.comment': {},
  'agent.github.pull': {},
  'agent.gate': { pattern: '' },
  'agent.note': { path: '' },
}

/** Every card on the board is this big, whatever it holds: a task is read as a chain, and
 *  a chain of things all the same size is the shape of the chain rather than the shape of
 *  its links. Laying one out by hand, leave a card's width between them. */
export const NODE_W = 96
export const NODE_H = 96

/** The next id nothing in this task has taken — `claude-1`, `interval-2`, `comment-1` — so
 *  an id says what it is as well as which one it is. The last part of the kind, not the
 *  second: everything that reaches GitHub has three, and `github-4` says nothing. */
export function freshId(task: Task, kind: TaskKind): string {
  const stem = kind.split('.').at(-1)
  const taken = new Set(task.nodes.map((node) => node.id))
  for (let n = 1; ; n++) if (!taken.has(`${stem}-${n}`)) return `${stem}-${n}`
}

/** A node, made the way the editor makes one: a fresh id, the kind's name, the fields the
 *  kind is born with, and its corner on the grid. */
export function makeNode(task: Task, kind: TaskKind, x: number, y: number): TaskNode {
  return {
    id: freshId(task, kind),
    kind,
    name: KIND_LABEL[kind],
    x: snap(x),
    y: snap(y),
    ...KIND_SEED[kind],
  } as TaskNode
}

File: daemon-go/internal/task/schema.go
// Package task is what a task is: a graph of triggers and agents, and the file it lives in. The
// shape only — reading and writing one is the codec's, walking one is the graph's.
//
// The TypeScript this is ported from writes a node as a discriminated union, one interface per
// kind. Go has no such thing, so a node is one struct wearing every field any kind can carry and
// a Kind that says which of them mean anything. The codec is where that stops being a loose bag:
// it reads only the fields the kind allows and writes only the fields the kind carried, so a
// field set on the wrong kind never reaches the file.
package task

import (
	"strconv"
	"strings"

	"github.com/broodmotherai/broodmother/daemon-go/internal/grid"
	"github.com/broodmotherai/broodmother/daemon-go/internal/utils"
)

const Extension = ".task"

func IsTaskPath(path string) bool { return utils.Extension(path) == "task" }

type Kind string

const (
	ManualTrigger   Kind = "trigger.manual"
	IntervalTrigger Kind = "trigger.interval"
	TimeTrigger     Kind = "trigger.time"
	FileTrigger     Kind = "trigger.file"

	GithubIssueTrigger   Kind = "trigger.github.issue"
	GithubPullTrigger    Kind = "trigger.github.pull"
	GithubMentionTrigger Kind = "trigger.github.mention"
	GithubCheckTrigger   Kind = "trigger.github.check"

	ClaudeAgent  Kind = "agent.claude"
	MuseAgent    Kind = "agent.muse"
	ShellAgent   Kind = "agent.shell"
	ApproveAgent Kind = "agent.approve"
	NotifyAgent  Kind = "agent.notify"
	HTTPAgent    Kind = "agent.http"

	GithubCommentAgent Kind = "agent.github.comment"
	GithubPullAgent    Kind = "agent.github.pull"

	GateAgent Kind = "agent.gate"
	NoteAgent Kind = "agent.note"
)

// Kinds is every kind a task can hold, in the order the add menu offers them.
var Kinds = []Kind{
	ManualTrigger, IntervalTrigger, TimeTrigger, FileTrigger,
	GithubIssueTrigger, GithubPullTrigger, GithubMentionTrigger, GithubCheckTrigger,
	ClaudeAgent, MuseAgent, ShellAgent, ApproveAgent, NotifyAgent, HTTPAgent,
	GithubCommentAgent, GithubPullAgent,
	GateAgent, NoteAgent,
}

// LegacyKinds are the kinds the editor no longer offers. A task written when it did still opens
// and runs; the menu just does not make any more of them.
var LegacyKinds = []Kind{GateAgent, NoteAgent}

// Weekday is a day a time trigger may fire on, as cron writes them.
type Weekday string

var Weekdays = []Weekday{"sun", "mon", "tue", "wed", "thu", "fri", "sat"}

func (d Weekday) valid() bool {
	for _, one := range Weekdays {
		if one == d {
			return true
		}
	}
	return false
}

// HTTPMethod is a verb a step may use. A closed list because a typo in a method is a request
// that fails at the far end for a reason nothing here would explain.
type HTTPMethod string

var HTTPMethods = []HTTPMethod{"GET", "POST", "PUT", "PATCH", "DELETE"}

func (m HTTPMethod) valid() bool {
	for _, one := range HTTPMethods {
		if one == m {
			return true
		}
	}
	return false
}

// Node is one step. Every kind carries an id, a name, a corner and the switch; the rest is the
// kind's, and which fields those are is written down once, in the codec.
type Node struct {
	ID   string
	Kind Kind
	Name string
	X    float64
	Y    float64
	// Off: the node does no work of its own and stands as a wire instead — what feeds it goes
	// straight on to what it feeds. A trigger switched off never fires, which is how a task keeps
	// its schedule on paper while you work on it. False is on, so every task written before the
	// switch reads as one with everything on.
	Off bool

	// Minutes is how often an interval trigger fires, how long a watch leaves GitHub alone, and
	// how long an errand may take. Unset on an errand is 5 — a step, not a day.
	Minutes *float64
	// At is a local time of day, HH:MM.
	At string
	// Days are the days a time trigger fires on. Unset is every day, which is what one written
	// before this said.
	Days []Weekday
	// Path is the file a trigger watches, or the note a run's output lands in.
	Path string
	// Append adds to the end of a note instead of rewriting it — how a recurring task keeps a log.
	Append *bool

	// Repo is `owner/name`. Unset is the remote of the checkout the task lives in.
	Repo *string
	// Query is a GitHub search, where one is wanted: `label:bug`, `review-requested:@me`. Unset
	// watches everything in the repository.
	Query *string
	// Branch is the branch whose checks are watched. Unset is the branch the checkout is on.
	Branch *string
	// Number is the issue or pull request commented on. Unset is whatever the trigger was about.
	Number *float64
	// Base is what a pull request is opened against. Unset is the repository's default branch.
	Base *string
	// Head is what it is opened from. Unset is the branch the checkout is on.
	Head *string
	// Title: the first line of the step's input is the title unless one is written here.
	Title *string
	Draft *bool

	Prompt string
	// Persona names a project persona whose PERSONA.md joins the agent's system prompt.
	Persona *string
	// Retries is how many more times to try after a failure. Unset is none.
	Retries *float64
	// Command is run by `sh -c` in the checkout, upstream output on stdin, stdout onward.
	Command string
	// Question is what an approval asks, where the node's name is not the whole question.
	Question *string

	URL string
	// Method unset is POST: the step has something to say, and saying it is what this is for.
	Method *HTTPMethod
	// Header is one header, written as it goes on the wire — `Authorization: Bearer …`.
	Header *string

	// Pattern: the branch continues only when the input matches — how "act on what the agent
	// flagged" is written.
	Pattern string
}

type Edge struct {
	From string
	To   string
}

type Task struct {
	Version int
	Nodes   []Node
	Edges   []Edge
}

// Empty is the task the app makes: the trigger that makes it runnable by hand, and nothing else
// yet. Unlike a canvas, which is fine with nothing on it, a task with no trigger is a task with
// no way to run.
func Empty() Task {
	return Task{
		Version: 1,
		Nodes:   []Node{{ID: "trigger", Kind: ManualTrigger, Name: ManualName, X: 80, Y: 120}},
		Edges:   []Edge{},
	}
}

func IsTrigger(kind Kind) bool { return strings.HasPrefix(string(kind), "trigger.") }

// IsGithub reports whether a kind reaches GitHub. The middle segment rather than the first,
// because both families have members there.
func IsGithub(kind Kind) bool { return strings.Contains(string(kind), ".github.") }

// IsGithubWatch reports whether a kind watches GitHub rather than writes to it — the ones that
// need a poll, and the ones that carry minutes.
func IsGithubWatch(kind Kind) bool { return IsGithub(kind) && IsTrigger(kind) }

// GithubKinds are the kinds that reach GitHub: a branch of their own in the add menu, under
// whichever family they belong to, and the ones the runtime needs a connection for.
var GithubKinds = githubKinds()

func githubKinds() []Kind {
	var found []Kind
	for _, kind := range Kinds {
		if IsGithub(kind) {
			found = append(found, kind)
		}
	}
	return found
}

// Fires reports whether a trigger will actually fire the task: wired into the graph, and
// switched on. Everything that keeps time or watches a source asks this one question.
func Fires(node Node, wired map[string]bool) bool { return wired[node.ID] && !node.Off }

// KindLabel is what each kind is called: the word in the add menu, and the name a node of it
// wears until it is given another.
var KindLabel = map[Kind]string{
	ManualTrigger:        ManualName,
	IntervalTrigger:      "Every N minutes",
	TimeTrigger:          "At a time",
	FileTrigger:          "When a file changes",
	ClaudeAgent:          "Claude Code",
	MuseAgent:            "Muse Code",
	ShellAgent:           "Command",
	ApproveAgent:         "Wait for approval",
	NotifyAgent:          "Notify me",
	HTTPAgent:            "Call a URL",
	GithubIssueTrigger:   "When an issue changes",
	GithubPullTrigger:    "When a pull request changes",
	GithubMentionTrigger: "When you are mentioned",
	GithubCheckTrigger:   "When checks change",
	GithubCommentAgent:   "Comment on GitHub",
	GithubPullAgent:      "Open a pull request",
	GateAgent:            "Only if",
	NoteAgent:            "Write a note",
}

// TriggerLabel is how a trigger reads in a sentence — the tasks page's word for it. Empty for an
// agent, which is not a thing that happens to you.
func TriggerLabel(node Node) string {
	switch node.Kind {
	case ManualTrigger:
		return "triggered manually"
	case IntervalTrigger:
		return "every " + number(node.Minutes) + " minute" + plural(node.Minutes)
	case TimeTrigger:
		return "at " + node.At + onDays(node.Days)
	case FileTrigger:
		return "when " + node.Path + " changes"
	case GithubIssueTrigger:
		return "when an issue changes in " + orThis(node.Repo, "repo") + asked(node.Query)
	case GithubPullTrigger:
		return "when a pull request changes in " + orThis(node.Repo, "repo") + asked(node.Query)
	case GithubMentionTrigger:
		return "when you are mentioned on GitHub"
	case GithubCheckTrigger:
		return "when checks change on " + orThis(node.Branch, "branch")
	default:
		return ""
	}
}

// number is a count as the template interpolating it would print it, which is JavaScript's
// number-to-string rather than Go's %v — 1.5 is "1.5" and 30 is "30", not "30.0".
func number(minutes *float64) string {
	if minutes == nil {
		return "undefined"
	}
	return strconv.FormatFloat(*minutes, 'g', -1, 64)
}

func plural(minutes *float64) string {
	if minutes != nil && *minutes == 1 {
		return ""
	}
	return "s"
}

func orThis(value *string, noun string) string {
	if value == nil {
		return "this " + noun
	}
	return *value
}

// onDays is the days a time trigger keeps to, where it keeps to any — read as part of the
// sentence the trigger is, since one that fires twice a week is not one that fires daily.
func onDays(days []Weekday) string {
	if len(days) == 0 {
		return ""
	}
	named := make([]string, len(days))
	for index, day := range days {
		named[index] = string(day)
	}
	return " on " + strings.Join(named, ", ")
}

// asked is the search a watch narrows itself with, where it has one — read as part of the
// sentence the trigger is, since a trigger watching one label is not watching the repository.
func asked(query *string) string {
	if query == nil || *query == "" {
		return ""
	}
	return " matching " + *query
}

// Every card on the board is this big, whatever it holds: a task is read as a chain, and a chain
// of things all the same size is the shape of the chain rather than the shape of its links.
// Laying one out by hand, leave a card's width between them.
const (
	NodeW = 96.0
	NodeH = 96.0
)

// FreshID is the next id nothing in this task has taken — `claude-1`, `interval-2`, `comment-1` —
// so an id says what it is as well as which one it is. The last part of the kind, not the second:
// everything that reaches GitHub has three, and `github-4` says nothing.
func FreshID(t Task, kind Kind) string {
	parts := strings.Split(string(kind), ".")
	stem := parts[len(parts)-1]
	taken := map[string]bool{}
	for _, node := range t.Nodes {
		taken[node.ID] = true
	}
	for n := 1; ; n++ {
		id := stem + "-" + strconv.Itoa(n)
		if !taken[id] {
			return id
		}
	}
}

// MakeNode makes a node the way the editor makes one: a fresh id, the kind's name, the fields
// the kind is born with, and its corner on the grid.
func MakeNode(t Task, kind Kind, x, y float64) Node {
	node := Node{
		ID:   FreshID(t, kind),
		Kind: kind,
		Name: KindLabel[kind],
		X:    grid.Snap(x),
		Y:    grid.Snap(y),
	}
	// The seed each kind is born with. Only three carry a value; the rest are born with the empty
	// string the struct already holds, which is what the TypeScript's empty partial amounts to.
	switch kind {
	case IntervalTrigger:
		minutes := 30.0
		node.Minutes = &minutes
	case TimeTrigger:
		node.At = "09:00"
	}
	return node
}
